qiskit/circuit/quantumcircuit.py

```diff
@@ -355,6 +356,23 @@ def from_instructions(
             circuit._append(instruction)
         return circuit
 
+    @property
+    def layout(self) -> Optional[TranspileLayout]:
```

Comment:
It's a good thing to have a different thing for "absence of value" - consumers should be forced to handle a missing value, not have things silently do the wrong thing. It helps the type checker or linter to scream at you, rather than have an error slip through to runtime. 

Ideally a compiler would force you to handle it, which is why languages with algebraic data types (Rust, many FPs like Haskell and OCaml, etc) expose an `Option<T>` type by default. Python doesn't have that so we're back to using a sentinel of a different type, but `mypy` and `pylint` should enforce the check in this form.

Reply:
"Empty values" like `[]` are useful when the output has some inherent variant size, but in this case it should be a class invariant that the `Layout` field of a `QuantumCircuit` contains mapping information for all the qubits in it. An empty layout wouldn't mean "do nothing", it would just be wrong, but in a way that static analysis can't tell you.

Reply:
I am not saying empty, though. I am saying `noop` / `null`. For instance `TrivialLayout` if no layout applies. That way users can request the _correct_ layout info directly regardless of whether the circuit was transpiled or not, removing checks which cost time and harm both performance and usability.

Reply:
Simply a general thought though, I have been using `metadata` lately and needing to write a lot of:
```python
if circuit.metadata is None:
    circuit.metadata = {}
circuit.metadata.update(...)
```
and worse... which feels a pointless when it could be set to an empty dict by default (i.e. in this other case empty would be the `noop` / `null` value).

Just applying the same general principle here of trying to make outputs have one and only one type whenever possible to save users time.

Reply:
Ok, I see what you mean more.  If the circuit hasn't been transpiled then it has no layout because there's no hardware for a layout to refer to, and assigning an identity map anyway would be harder for consumers because you wouldn't be able to tell the difference between "layout that happens to be trivial" and "not laid out".  For metadata, that does look weird to me - I don't know why the initialiser doesn't construct an empty dictionary.

Reply:
This is a good idea in principal. But how would returning something other than `None` work practically? If this were the general case
```python
if circuit.metadata is None:
    circuit.metadata = {}
circuit.metadata.update(...)
```
then returning `{}` would be better design. `TrivialLayout`
https://github.com/Qiskit/qiskit-terra/blob/8af40458387200981c28ea1b803437cb9b35bd95/qiskit/transpiler/passes/layout/trivial_layout.py#L20
is a transpiler pass. Maybe you mean the thing that a `TrivialLayout` calls: `Layout.generate_trivial_layout` ? But I don't think this could be correct, either. `TranspileLayout` isn't a layout, but rather an object that contains a `Layout` and other data as well. How often would one want to check if the `QuantumCircuit` has an attached `TranspileLayout`, and if not, do nothing? In this case doing the minimum initialization would be more costly than returning and checking for `None`?
Practical use cases would make the case for something other than `None` stronger.

Reply:
Oh! I did not recall that `TrivialLayout` was already taken. I meant literally a _trivial layout_, as in the identity permutation (if you would indulge me forgetting about qubit objects for a minute), which for $n$ qubits is simply $(0, 1, 2, ..., n-1)$. In other words return the information representing that _nothing_ has changed: since nothing has really changed, either because transpilation did not occur, or because it did not swap any qubits —users/developers should not be forced to care if they are developing something on top and are handed down a circuit whose origins are unknown.

My point was to try to eliminate some of the "statefulness" in `QuantumCircuit`, so that users dealing with one such object do not need to care what its history is, but rather what it says right now. In this case, I don't care if the circuit was transpiled or not, I care that it has certain gates and certain layout (etc), regardless of where it came from. I often find that this improves usability quite a bit, and simplifies automation of other tasks on top.

This happens for instance in the different implementations of `BaseEstimator`, where we need to make observables crafted pre-transpilation fit the layout of our circuit after transpilation, and therefore we need to somehow extract information about the layout and pass it along to another function to handle observables. We could, nonetheless, `skip_transpilation` in which case there will not be a layout to report, and we will need to introduce conditional branches to account for this fact. The alternative is that we always have something to report, avoiding conditional branching. We simply retrieve a `noop` layout, pass it along, and because it is `noop` it will not change the observables. No branching needed and the code stays simpler.

I do accept @jakelishman comments though, so please correct me if I am forgetting about something.

Not sure I fully understand your comment about metadata @jlapeyre , but I opened an issue to track it. Maybe it is better to transition that conversation there: https://github.com/Qiskit/qiskit-terra/issues/9498

Let me try to answer your question on practicality, though, and forgive me if what I say is too obvious.

The idea is essentially the same. If I build a function that gets an unknown circuit as an argument, and such function updates the metadata, I as a developer would like to do it as seamlessly as possible:
```python
def do_magic(circuit):
    ...
    circuit.metadata.update({...})
    return ...
```
Now, since I don't know where the circuit comes from, I have no guarantee that I can call that line of code directly if the default is `None`, and I am forced to type check.

Someone who wants to check if there is some metadata can still do so by:
```python
if circuit.metadata:
    ...
```
which evaluates to `False` if `{}`.

To set this up we would only need `__init__` to set `self.metadata = metadata or {}`. Is this what you meant?

Reply:
I meant that the case for returning an empty collection for `metadata` is strong. And wondered if there is an equally strong argument for `_layout`. If I understand @jakelishman 's comment, it argues against doing this for `_layout`. The argument is reasonable at least for the case where one needs to know whether a layout has been done for a reason, eg for particular hardware. I don't think it makes sense to switch to returning a trivial (small t) layout unless you have an explicit need for it. And even then, you might want another method that means "return the layout or trivial layout if there is none".